File: JavaScriptPractice/Leetcode567.js
function dictEqual(d1,d2)
{
    for(let [key,val] of d1)
    {
        if(d2.get(key)!=val){return false;}
    }
    return true;
}
var checkInclusion = function(s1, s2) {
    let s1len = s1.length;
    let s2len = s2.length;
    if(s1len>s2len){return false;}
    let s1Map = new Map();
    let tempMap = new Map();
    for (let c='a';c<='z';c=String.fromCharCode(c.charCodeAt(0)+1))
    {
        s1Map.set(c,0); 
    }
    for(let c of s1){s1Map.set(c,s1Map.get(c)+1);}
    for (let c='a';c<='z'; c=String.fromCharCode(c.charCodeAt(0)+1))
    {
        tempMap.set(c,0);
    }

    for(let c of s2.substring(0,s1len)){tempMap.set(c,tempMap.get(c)+1);}
    let fast = s1len-1;
    for(let i=0;i<s2len-s1len;i++)
    {
        // for(let [key,val] of tempMap){console.log(key,val);}
        if(dictEqual(tempMap,s1Map)){return true;}
        fast++;
        if(fast<s2len)
        {
            tempMap.set(s2[fast],tempMap.get(s2[fast])+1);
            tempMap.set(s2[i],tempMap.get(s2[i])-1);
        }
    }
    if(dictEqual(tempMap,s1Map)){return true;}else{return false;}

};

File: CSharpPractice/LeetcodePractice/Leetcode567.cs
public class Solution567 {
    private bool dictEqual(Dictionary<char,int> d1, Dictionary<char,int> d2)
    {
        foreach(var pair in d1)
        {
            if(pair.Value != d2[pair.Key]){return false;}
            
        }
        return true;

    }
    public bool CheckInclusion(string s1, string s2) 
    {
        int s1len = s1.Length;
        int s2len = s2.Length;
        if(s1len>s2len){return false;}
        Dictionary<char,int> s1Map = new Dictionary<char,int>();
        for(char c='a';c<='z';c++){s1Map[c]=0;}
        foreach(char c in s1){s1Map[c]++;}
        Dictionary<char,int> tempMap = new Dictionary<char,int>();
        for(char c='a';c<='z';c++){tempMap[c]=0;}
        foreach(char c in s2.Substring(0,s1len)){tempMap[c]++;}
        
        int slow = 0;
        int fast=s1len-1;

        for(int i=0;i<s2len-s1len;i++)
        {
            foreach(var kvp in tempMap){Console.WriteLine($"{kvp.Key},{kvp.Value}");}
            if(dictEqual(s1Map,tempMap)){return true;}
            fast++;
            if(fast<=s2len-1)
            {
                tempMap[s2[fast]]++;
                tempMap[s2[slow]]--;
                slow++;
            }
        }
        if(dictEqual(s1Map,tempMap)){return true;}else{return false;}
    }
}